### ISSUE TITLE
[Bug]: AC0032 - False Positive for DataTransfer and CopyFields

### ISSUE THREAD

**Issue by pri-kise** [ALCops/Analyzers#465]
### Analyzer

ApplicationCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

AC0032

### Description

The warning is wrong when the table is changed with DataTransfer. 
DataTransfer checks the license - therefore we need to enable indirect permissions.
> Permission 'tabledata Purch. Cr. Memo Line' is declared but this object performs no database operations on table 'Purch. Cr. Memo Line'.

### AL code to reproduce

```al
codeunit 50000 "PTE Upgrade"
{
    //* https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-upgrading-extensions
    Access = Internal;
    Permissions =
        tabledata "Purch. Cr. Memo Line" = rm;
    Subtype = Upgrade;

    procedure UpgradePurchCrMemoLine(OldValue: Code[20]; NewValue: Code[20])
    var
        PurchCrMemoLine: Record "Purch. Cr. Memo Line";
        PurchCrMemoLineDataTransfer: DataTransfer;
    begin
        PurchCrMemoLineDataTransfer.SetTables(Database::"Purch. Cr. Memo Line", Database::"Purch. Cr. Memo Line");
        PurchCrMemoLineDataTransfer.AddSourceFilter(PurchCrMemoLine.FieldNo("PTE Custom Code"), '=%1', OldValue);
        PurchCrMemoLineDataTransfer.AddConstantValue(NewValue, PurchCrMemoLine.FieldNo("PTE Custom Code"));
        PurchCrMemoLineDataTransfer.CopyFields();
    end;
}
```

**Comment by Arthurvdv** [comment]
## Investigation & recommendation

Confirmed. The `DataTransfer` type is completely invisible to AC0031/AC0032 today, so any object whose only access to a table goes through `DataTransfer` gets its `Permissions` entry flagged as unused. Related to #466 (missing read operator), but this one needs a small dedicated detector rather than a new `MethodOperationMap` entry, because the table is not the receiver of the call.

### How DataTransfer touches the database

From the SDK (`Microsoft.Dynamics.Nav.CodeAnalysis/.../Symbols/DataTransferClassTypeSymbol.cs`) and the [DataTransfer docs](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/devenv-data-transfer), the type is a builder:

| Method | DB effect |
|---|---|
| `SetTables(SourceTableId, DestinationTableId)` | none — names the two tables (integers, in practice `Database::"X"` literals) |
| `AddFieldValue`, `AddConstantValue`, `AddJoin`, `AddSourceFilter`, `AddDestinationFilter`, `UpdateAuditFields` | none — in-memory configuration (field numbers / values / flags) |
| **`CopyFields()`** | one bulk SQL `UPDATE`: **Read** on source, **Modify** on destination (no OnModify triggers) |
| **`CopyRows()`** | one bulk SQL `INSERT ... SELECT`: **Read** on source, **Insert** on destination (no OnInsert triggers) |

When source and destination are the same table (the repro, and the documented `ModifyAll` replacement pattern) this collapses to `rm` for `CopyFields` and `ri` for `CopyRows` on one table — which is exactly what the reporter declared. DataTransfer can only run in upgrade/install code, can't target temporary, system, virtual, audited or obsolete tables, and no Microsoft cop models it (nothing in CodeCop/AppSourceCop/PTECop references `DataTransfer`).

### Options considered

- **A — treat like `RecordRef` and bail out of the whole object.** Simple, but upgrade codeunits are precisely where stale `Permissions` accumulate, so we'd lose AC0032 exactly where it's useful.
- **B — resolve the tables from `SetTables` and attribute `CopyFields`/`CopyRows` to them.** Gives real answers for the common case (`Database::X` literals).

**Decision: hybrid.** Do B when both `SetTables` arguments resolve statically; fall back to the existing whole-object bailout (A) when they don't.

### Decisions already taken (don't re-open)

1. `CopyFields` = `r`(source) + `m`(destination); `CopyRows` = `r`(source) + `i`(destination). Full char granularity — partial-unused (`rimd` declared, only `rm` needed) is still reported.
2. AC0031 gets the symmetric behaviour (missing `r`/`m`/`i` for a DataTransfer is reported). Diagnostic is anchored on the `CopyFields()` / `CopyRows()` call, consistent with anchoring on `Modify()` / `Insert()`.
3. Resolution scope is **the same method/trigger body**: pair each executor call with the last `SetTables` on the same DataTransfer variable that textually precedes it. The variable may be local, global or parameter — but the `SetTables` must be in the same body.
4. If an executor call has no resolvable `SetTables` in the same body (set in another procedure, `var` parameter, argument is a variable/parameter/expression rather than `Database::X`): **AC0032 → whole-object bailout** (same mechanism as `RecordRef`, #420), **AC0031 → silent** (cannot name a table).
5. Only `CopyFields` and `CopyRows` are operations; the other seven methods are ignored. Do **not** add them to `MethodOperationMap` — that map is keyed on record receivers and is mirrored by `RecordMethodClassification`; use a separate small set (e.g. `DataTransferOperations`: `CopyFields → (Read, Modify)`, `CopyRows → (Read, Insert)`).

### Resolving `Database::"X"`

`Database::"Purch. Cr. Memo Line"` binds to `BoundApplicationObjectAccess`, exposed as `IApplicationObjectAccess` with `.ApplicationObjectTypeSymbol` (an `ITableTypeSymbol` for `Database::`). It reports `OperationKind.FieldAccess`, so the SDK's `GetSymbol()` throws — use the existing `OperationSafeExtensions.GetSymbolSafe()` in `src/ALCops.Common/Extensions/OperationExtensions.cs`, which already handles exactly this case. Unwrap `IConversionExpression` around the argument first (same pattern as `PlatformCop/RecordGetProcedureArguments.cs`). Anything that isn't an `IApplicationObjectAccess` → unresolved → rule 4 above. Apply the usual filters afterwards (`IsTemporary()`, `IsSystemTable` per rule).

### Implementation sketch

**Common (`src/ALCops.Common/Permissions/`)**
- `RequiredPermissionDetector.GetFromDataTransfer(IInvocationExpression executor, SemanticModel model, bool includeSystemTables)` → up to 2 `RequiredPermission`s (`VariableType` = the table type itself, like `TryGetPermissionForType` does for bare self calls), plus an "unresolvable" signal so callers can distinguish "no permission needed" from "unknown table".
- Helper to find the paired `SetTables`: from `executor.Syntax` go to `FirstAncestorOrSelf<MethodOrTriggerDeclarationSyntax>()`, enumerate `InvocationExpressionSyntax` whose expression is a `MemberAccessExpressionSyntax` named `SetTables` with the same receiver identifier, keep the last one with `SpanStart < executor.Syntax.SpanStart`, then `model.GetOperation(...)` → `IInvocationExpression.Arguments[0/1].Value`.

**AC0031 (`TableDataAccessRequiresPermissions`)**
- In `AnalyzeInvocation`, next to the existing `TryGetFromInvocation`: if `invocation.TargetMethod.MethodKind == BuiltInMethod`, `invocation.Instance?.Type.NavTypeKind == EnumProvider.NavTypeKind.DataTransfer` (already in `EnumProvider`) and the name is `CopyFields`/`CopyRows`, call the new detector and run each result through `PermissionResolver.IsCovered`. SemanticModel via `ctx.Compilation.GetSemanticModel(invocation.Syntax.SyntaxTree)` (or `ctx.SemanticModel` where the context exposes it — see c702b4b for LC0089).

**AC0032 (`TableDataAccessUnusedPermissions`)**
- Track DataTransfer-typed locals/globals/params/named returns by name next to the existing RecordRef name sets (`NavTypeKind.DataTransfer`), same shadowing rules.
- Extend `HasPossibleDbInvocation` to also match `CopyFields`/`CopyRows` (they're not in `MethodOperationMap`, so the pre-filter would otherwise skip the whole method).
- In `TryGetPermissionFromDbAccess`, when the method name is an executor and the receiver resolves to a DataTransfer variable: call the detector; add the resulting permissions to `requiredPermissions`, or set the bailout flag when unresolvable. Paren-less `dt.CopyFields;` is legal AL, so keep the `MemberAccessExpressionSyntax` branch working.

### Test cases

`TableDataAccessUnusedPermissions/`
- `NoDiagnostic/DataTransferCopyFieldsSameTable.al` — the repro (`rm` declared, same-table `CopyFields`).
- `NoDiagnostic/DataTransferCopyRowsTwoTables.al` — `Source = r`, `Destination = i`.
- `NoDiagnostic/DataTransferCopyFieldsTwoTables.al` — `Source = r`, `Destination = m`.
- `NoDiagnostic/DataTransferGlobalVariable.al` — DataTransfer as a global var, `SetTables` + `CopyFields` in the same procedure.
- `NoDiagnostic/DataTransferUnresolvableSetTables.al` — `SetTables(TableId, TableId)` with an integer parameter, plus an unrelated unused entry → bailout, no diagnostic.
- `NoDiagnostic/DataTransferSetTablesInOtherProcedure.al` — bailout path.
- `NoDiagnostic/DataTransferWithoutParentheses.al` — `dt.CopyFields;`.
- `HasDiagnostic/DataTransferPartialUnused.al` — `rimd` declared, same-table `CopyFields` → `id` unused (`UsedChars = "rm"`).
- `HasDiagnostic/DataTransferCopyRowsDestinationPartialUnused.al` — `Destination = rimd`, two-table `CopyRows` → `rmd` unused on destination (`UsedChars = "i"`).
- `HasDiagnostic/DataTransferOnlyBuilderCalls.al` — `SetTables` + `AddSourceFilter` but no executor → entry still unused.

`TableDataAccessRequiresPermissions/`
- `HasDiagnostic/DataTransferCopyFieldsMissingModify.al` — `= r` declared, same-table `CopyFields` → missing `m` at the `CopyFields()` call.
- `HasDiagnostic/DataTransferCopyRowsMissingInsert.al` — destination lacks `i`.
- `HasDiagnostic/DataTransferMissingReadOnSource.al` — two tables, source has no entry.
- `NoDiagnostic/DataTransferCopyFieldsWithPermissions.al`, `NoDiagnostic/DataTransferUnresolvable.al` (integer variable → silent), `NoDiagnostic/DataTransferInherentPermissions.al`.

### Docs / instruction files

- `.github/instructions/ac0031-…` and `ac0032-…`: new detector, design-decision rows, a Known-limitation row ("`SetTables` in a different procedure than `CopyFields` → bailout"), test counts.
- `.github/instructions/common-library.instructions.md`: mention `GetFromDataTransfer` and the separate `DataTransferOperations` set.
- AC0031/AC0032 rule pages on alcops.dev: state that `DataTransfer.CopyFields`/`CopyRows` count as `r`+`m` / `r`+`i` on the tables named in `SetTables`, and that a non-literal `SetTables` disables AC0032 for the object.

**Comment by Arthurvdv** [comment]
Fixed in #491 (354bc54): `DataTransfer.CopyFields`/`CopyRows` now count as `r`+`m` / `r`+`i` on the tables named in `SetTables` for both AC0031 and AC0032. Docs: ALCops/alcops.dev#157. Follow-up for the AC0031 CodeFix multi-table case: #492.

**Comment by Arthurvdv** [comment]
I've published a new beta release, `v1.2.0-beta.1`, which should resolve this issue.

Could you verify for me if it's fixed on your end? Would be great to hear whether it's resolved or if you're still running into problems.

If you're using the [ALCops VS Code extension](https://marketplace.visualstudio.com/items?itemName=Arthurvdv.alcops), you can switch to the beta channel in the extension settings. To force an update check, run the `ALCops: Check for Updates` command from the command palette.

**Comment by Arthurvdv** [comment]
This should now be resolved with the release of `v1.2.0` of ALCops.

When you have a chance, could you update and verify whether the issue is fixed?

If everything is working as expected, feel free to close this issue. If you still run into the problem, please leave the issue open and share any additional context, code samples, or reproduction steps that might help me investigate further.
